utils/pstor-usage/src/resources.rs

```diff
@@ -0,0 +1,112 @@
+#[async_trait::async_trait]
+pub(crate) trait ResourceMgr: Send + Sync {
+    type Output: ResourceDelete;
+    /// Create `count` resources
+    async fn create(&self, client: ApiClient, count: u32) -> anyhow::Result<Self::Output>;
+    /// Delete the created resources
+    async fn delete(&self, client: ApiClient, created: Self::Output) -> anyhow::Result<()>;
+    /// Prepare a sample from the given data points
+    fn prepare_sample(&self, points: Vec<u64>) -> Box<dyn ResourceSample>;
+}
+
+/// Resources that are deletable and should be deleted
+#[async_trait::async_trait]
+pub(crate) trait ResourceDelete: Send + Sync + Clone {
+    async fn delete(&self, client: ApiClient) -> anyhow::Result<()>;
+}
+
+#[async_trait::async_trait]
+impl<T: ResourceDelete> ResourceDelete for Vec<T> {
+    async fn delete(&self, client: ApiClient) -> anyhow::Result<()> {
+        for e in self.iter() {
+            e.delete(client.clone()).await?;
+        }
+        Ok(())
+    }
+}
+
+/// Sample `count` data points from a given `ResourceMgr`
+#[async_trait::async_trait]
+pub(crate) trait Sampler {
+    async fn sample<T: ResourceMgr>(
+        &self,
+        client: ApiClient,
+        count: u32,
+        resource_mgr: &T,
+    ) -> anyhow::Result<(Vec<T::Output>, ResourceSamples)>;
+}
+
+/// A sample of data points, if you will
+pub(crate) trait ResourceSample {
+    /// Get a reference to the data points
+    fn points(&self) -> &Vec<u64>;
+    /// Get a mutable reference to the data points
+    fn points_mut(&mut self) -> &mut Vec<u64>;
+    /// Accrue the data points with the given ones
+    fn accrue_points(&mut self, add: &[u64]) {
+        for (i, p) in self.points_mut().iter_mut().enumerate() {
+            if let Some(p2) = add.get(i) {
+                *p += *p2;
+            }
+        }
+        if add.len() > self.points().len() {
+            let add = &add[self.points().len() .. add.len()];
+            self.points_mut().extend(add);
+        }
+    }
+    /// Get the name of the data points
+    fn name(&self) -> String;
+    /// Format a data point into a String
+    fn format_point(&self, point: u64) -> String;
+}
+
+/// A helper collection of samples
+pub(crate) struct ResourceSamples {
+    inner: Vec<Box<dyn ResourceSample>>,
+}
+
+impl ResourceSamples {
+    /// Return a new `Self` with the starting samples
+    pub(crate) fn new(samples: Vec<Box<dyn ResourceSample>>) -> Self {
+        Self { inner: samples }
+    }
+    /// Move out the samples
+    pub(crate) fn into_inner(self) -> Vec<Box<dyn ResourceSample>> {
+        self.inner
+    }
+    /// Get reference to the samples
+    pub(crate) fn inner(&self) -> &Vec<Box<dyn ResourceSample>> {
+        &self.inner
+    }
+    /// Get length of the samples
+    pub(crate) fn len(&self) -> usize {
+        self.inner.len()
+    }
+    /// Add more samples
+    pub(crate) fn extend(&mut self, add: Vec<Box<dyn ResourceSample>>) {
+        self.inner.reverse();
```

Comment:
Why do we need to reverse it and then reverse it again at the end of the function?

Reply:
It's to keep the same order, we could probably do it in a simpler way though.. for now I've added a comment.